internal/sync/lifecycle.go

```diff
@@ -20,24 +20,105 @@
 
 package sync
 
+import (
+	"sync"
+
+	"go.uber.org/atomic"
+)
+
+// LifecycleState represents `states` that a lifecycle object can be in.
+type LifecycleState int
+
+const (
+	// Idle indicates the Lifecycle hasn't been operated on yet.
+	Idle LifecycleState = iota
+
+	// Starting indicates that the Lifecycle has begun it's "start" command
+	// but hasn't finished yet.
+	Starting
+
+	// Running indicates that the Lifecycle has finished starting and is
+	// available.
+	Running
+
+	// Stopping indicates that the Lifecycle 'stop' method has been called
+	// but hasn't finished yet.
+	Stopping
+
+	// Stopped indicates that the Lifecycle has been stopped.
+	Stopped
+
+	// Errored indicates that the Lifecycle experienced an error and we can't
+	// reasonably determine what state the lifecycle is in.
+	Errored
+)
+
 // LifecycleOnce is a helper for implementing transport.Lifecycles
 // with similar behavior.
 type LifecycleOnce struct {
-	start Once
-	stop  Once
+	lock sync.Mutex
+
+	state     atomic.Int32
+	startOnce sync.Once
+	startErr  error
+	stopOnce  sync.Once
+	stopErr   error
```

Comment:
I believe you only need a single atomic for the state. The atomic state is your lock. I am not sure about the utility of storing the errors?